### PR TITLE
Optimize radix tree related modules

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Use terminalNodeMap instead of stateNodeMap (target: state node -> terminal radix node)
- Optimize _getRadixNodeForReading() using terminalNodeMap 
- Optimize copyFrom()
- Immediately update state info (by calling updateStateInfo()) from StateNode.deleteChild()
- Keep tree info in RadixNode
- Deprecate hexLabelCache and RadixChildMap
- Optimize deleteStateTreeVersion() using deleteRadixTree() 
- Fix and simplify updateStateInfoForAllRootPaths()

Related issues: 
https://github.com/ainblockchain/ain-blockchain/issues/519
https://github.com/ainblockchain/ain-blockchain/issues/517

### CODE REVIEW

File: db/radix-tree.js [hunk 1]
```diff
@@ -244,8 +202,9 @@ class RadixTree {
       return false;
     }
     node.resetStateNode();
+    let nodeToUpdateProofHash = node;
     if (node.numChildren() === 1) {
-      return this._mergeToChild(node);
+      nodeToUpdateProofHash = this._mergeToChild(node, updateProofHash);
```
Comment (liayoo) [inline db/radix-tree.js:207]: should the second arg be removed as _mergeToChild() only needs 1 paramter?

File: db/radix-tree.js [hunk 2]
```diff
@@ -257,68 +216,98 @@ class RadixTree {
       const parent = node.getParent();
       parent.deleteChild(node.getLabelRadix());
       if (parent.numChildren() === 1 && !parent.hasStateNode() && parent.hasParent()) {
-        return this._mergeToChild(parent);
+        nodeToUpdateProofHash = this._mergeToChild(parent, updateProofHash);
```
Comment (liayoo) [inline db/radix-tree.js:219]: ditto

File: db/radix-tree.js [hunk 3]
```diff
@@ -137,63 +119,40 @@ class RadixTree {
     return curNode;
   }
 
-  _getFromTree(stateLabel) {
-    const hexLabel = this._toHexLabel(stateLabel);
-    const node = this._getRadixNodeForReading(hexLabel);
-    if (node === null) {
+  // NOTE(platfowner): Use hash map instead of radix tree as it faster.
+  get(stateLabel) {
+    const LOG_HEADER = '_getFromMap';
```
Comment (liayoo) [inline db/radix-tree.js:124]: _getFromMap -> get?

File: db/radix-tree.js [hunk 4]
```diff
@@ -331,20 +320,22 @@ class RadixTree {
...
-  }
-
   copyFrom(radixTree, newParentStateNode) {
-    this._copyTreeFrom(radixTree, newParentStateNode);
-    this._copyMapFrom(radixTree);
+    const LOG_HEADER = 'copyFrom';
+
+    const terminalNodeMap = new Map();
+    this.root.copyFrom(radixTree.root, newParentStateNode, terminalNodeMap);
+    // Keep the insertion order.
```
Comment (liayoo) [inline db/radix-tree.js:328]: nice! 🙌

File: db/radix-tree.js (hunk 1 above)
Comment (platfowner) [reply db/radix-tree.js]: Fixed. Thanks!

File: db/radix-tree.js (hunk 2 above)
Comment (platfowner) [reply db/radix-tree.js]: ditto

File: db/radix-tree.js (hunk 3 above)
Comment (platfowner) [reply db/radix-tree.js]: Just fixed. Thanks!

**Comment by platfowner** [comment]
Addressed the comments so far~ PTAL @liayoo

File: db/radix-tree.js [hunk 5]
```diff
@@ -221,14 +180,13 @@ class RadixTree {
...
     const newChildLabelSuffix = labelSuffix + childLabelRadix + childLabelSuffix;
     parent.setChild(labelRadix, newChildLabelSuffix, child);
-    return true;
+    return parent;
   }
 
-  _deleteFromTree(stateLabel) {
-    const LOG_HEADER = '_deleteFromTree';
+  delete(stateLabel, updateProofHash = false) {
+    const LOG_HEADER = '_delete';
```
Comment (liayoo) [inline db/radix-tree.js:187]: _delete -> delete

File: db/state-util.js [hunk 6]
```diff
@@ -612,55 +635,35 @@ function equalStateTrees(node1, node2) {
...
-          `at: ${new Error().stack}.`);
-      needToUpdate = true;
-      break;
-    }
-    curNode = child;
-  }
-  return updateStateInfoForAllRootPathsRecursive(curNode, needToUpdate, null);
+  const childNode = curNode.getChild(updatedChildLabel);
+  return updateStateInfoForAllRootPathsRecursive(
+      curNode, updatedChildLabel, isEmptyNode(childNode));
```
Comment (liayoo) [inline db/state-util.js:666]: question) is childNode ever empty here?

**Review by liayoo — APPROVED** [review]
LGTM with a comment & a question. Thanks!

File: db/state-util.js (hunk 6 above)
Comment (platfowner) [reply db/state-util.js:666]: yes, it's the case we set value null for deletion.
fyi, integration test cases prefixed 'updateStateInfoForAllRootPaths on empty node' cover this case.

**Comment by platfowner** [comment]
Thanks for the fast review. Now merging..

**Pull request merged**